compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -741,9 +773,55 @@ impl<'a> Parser<'a> {
             err.span_label(sp, label_exp);
             err.span_label(self.token.span, "unexpected token");
         }
+
+        // Check for misspelled keywords if there are no suggestions added to the diagnostic.
+        if err.suggestions.as_ref().is_ok_and(|code_suggestions| code_suggestions.is_empty()) {
+            self.check_for_misspelled_kw(&mut err, &expected);
+        }
         Err(err)
     }
 
+    /// Checks if the current token or the previous token are misspelled keywords
+    /// and adds a helpful suggestion.
+    fn check_for_misspelled_kw(&self, err: &mut Diag<'_>, expected: &[TokenType]) {
+        let Some((curr_ident, _)) = self.token.ident() else {
+            return;
+        };
+        let expected_tokens: &[TokenType] =
+            expected.len().checked_sub(10).map_or(&expected, |index| &expected[index..]);
```

Comment:
why we only considering last 10 expected tokens, is there any bad case we want to avoid?

Reply:
`expected` is obtained from `Parser`'s `expected_tokens` field. This list of expected tokens can be arbitrarily long depending on the size of the program because a token is added to the list every time the `check_*` functions are invoked inside the parser. 

But the actual expected keyword is located in the tail end of the list. For example, the parser adds the `where` keyword along with a few other tokens to the end of the expected tokens list after parsing a function's return type. 

So I arbitrarily chose the last ten tokens to make it efficient for the edit distance algorithm to process the entire list of keywords.